### ISSUE TITLE
prefor does not update seen

### ISSUE THREAD

**Issue by DrChainsaw** [FluxML/Flux.jl#803]
I'm guessing that IdSet is there to avoid infinite recursion, but nothing is ever put into it.

For example:
```julia
using Flux
mutable struct A a end
Flux.@treelike A

trouble = A(nothing)
trouble.a = trouble

params(trouble)
ERROR: StackOverflowError:
```

Happens on 0.8.3, but source on master looks the same to me.

**Comment by MikeInnes** [comment]
Yes, I think also to avoid collecting parameters twice from layers that are used multiple times, which would be another interesting test case to come up with.

Would be happy to give pointers if you're interested in submitting a PR for this.

**Comment by DrChainsaw** [comment]
Thanks,

Sorry for not sending a PR right away. I'm new to the language so I wanted to make sure there was no other intention.

PR sent: #805

**Comment by MikeInnes** [comment]
Closed by #805.
